packages/upload-api/src/blob/add.js

```diff
@@ -146,37 +152,66 @@ async function allocate({ context, blob, space, cause }) {
   const { router } = context
   const digest = Digest.decode(blob.digest)
 
-  const candidate = await router.selectStorageProvider(digest, blob.size)
-  if (candidate.error) {
-    return candidate
-  }
+  let provider, task, receipt
+  /** @type {Principal[]} */
+  let exclude = []
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const candidate = await router.selectStorageProvider(digest, blob.size, {
+      exclude,
+    })
+    if (candidate.error) {
+      return candidate
+    }
 
-  const cap = Blob.allocate.create({
-    with: candidate.ok.did(),
-    nb: {
-      blob: {
-        digest: blob.digest,
-        size: blob.size,
+    const cap = Blob.allocate.create({
+      with: candidate.ok.did(),
+      nb: {
+        blob: {
+          digest: blob.digest,
+          size: blob.size,
+        },
+        space: DID.parse(space),
+        cause,
       },
-      space: DID.parse(space),
-      cause,
-    },
-  })
+    })
 
-  const configure = await router.configureInvocation(candidate.ok, cap, {
-    expiration: Infinity,
-  })
-  if (configure.error) {
-    return configure
-  }
+    const configure = await router.configureInvocation(candidate.ok, cap, {
+      expiration: Infinity,
+      channel: { fetch: fetchWithTimeout(allocateExecTimeout) },
+    })
+    if (configure.error) {
+      return configure
+    }
 
-  const task = await configure.ok.invocation.delegate()
-  const receipt = await configure.ok.invocation.execute(configure.ok.connection)
+    try {
+      provider = candidate.ok
+      task = await configure.ok.invocation.delegate()
+      receipt = await configure.ok.invocation.execute(configure.ok.connection)
+      if (receipt.out.error) {
+        throw new Error(`allocation failed: ${receipt.out.error.message}`, {
+          cause: receipt.out.error,
+        })
+      }
+      break
+    } catch (err) {
+      console.error(
+        `executing ${
+          Blob.allocate.can
+        } on ${provider?.did()} for blob ${base58btc.encode(blob.digest)} (${
+          blob.size
+        } bytes)`,
+        err
+      )
+      exclude.push(candidate.ok)
+      // loop until we run out of candidates
+    }
+  }
 
   // record the invocation and the receipt, so we can retrieve it later when we
   // get a http/put receipt in ucan/conclude
   const message = await Message.build({
-    invocations: [configure.ok.invocation],
+    invocations: [task],
```

Comment:
task appears to be configure.ok.invocation.delegate() now -- not just configure.ok.invocation -- is this correct?